packages/core/src/actions/contracts/readContracts.ts

```diff
@@ -8,37 +14,116 @@ import {
   ContractMethodRevertedError,
   ContractResultDecodeError,
 } from '../../errors'
+import {
+  GetArgs,
+  GetResult as ReadContractsContractResult,
+} from '../../types/utils'
 import { logWarn } from '../../utils'
 import { getProvider } from '../providers'
 import { multicall } from './multicall'
-import { ReadContractConfig, readContract } from './readContract'
+import { readContract } from './readContract'
+
+type ReadContractsContractConfig<
+  TAbi extends Abi | readonly unknown[] = Abi,
+  TFunctionName extends string = string,
+  TFunction extends AbiFunction & { type: 'function' } = TAbi extends Abi
+    ? ExtractAbiFunction<TAbi, TFunctionName>
+    : never,
+> = {
+  /** Contract address */
+  addressOrName: Address
+  /** Chain id to use for provider */
+  chainId?: number
+  /** Contract ABI */
+  contractInterface: TAbi
+  /** Function to invoke on the contract */
+  functionName: [TFunctionName] extends [never] ? string : TFunctionName
+} & GetArgs<TAbi, TFunction>
 
-export type ReadContractsContract = {
-  addressOrName: ReadContractConfig['addressOrName']
-  args?: ReadContractConfig['args']
-  chainId?: ReadContractConfig['chainId']
-  contractInterface: ReadContractConfig['contractInterface']
-  functionName: ReadContractConfig['functionName']
+type GetConfig<T> = T extends {
+  contractInterface: infer TAbi extends Abi
+  functionName: infer TFunctionName extends string
 }
+  ? ReadContractsContractConfig<
+      TAbi,
+      ExtractAbiFunctionNames<TAbi, 'view' | 'pure'>,
+      ExtractAbiFunction<TAbi, TFunctionName>
+    >
+  : ReadContractsContractConfig
+
+type GetResult<T> = T extends {
+  contractInterface: infer TAbi extends Abi
+  functionName: infer TFunctionName extends string
+}
+  ? ReadContractsContractResult<TAbi, TFunctionName>
+  : ReadContractsContractResult
+
+type MAXIMUM_DEPTH = 20
+type ContractsConfig<
+  TContracts extends unknown[],
+  Result extends any[] = [],
+  Depth extends ReadonlyArray<number> = [],
+> = Depth['length'] extends MAXIMUM_DEPTH
+  ? ReadContractsContractConfig[]
+  : TContracts extends []
+  ? []
+  : TContracts extends [infer Head]
+  ? [...Result, GetConfig<Head>]
+  : TContracts extends [infer Head, ...infer Tail]
+  ? ContractsConfig<[...Tail], [...Result, GetConfig<Head>], [...Depth, 1]>
+  : unknown[] extends TContracts
+  ? TContracts
+  : TContracts extends ReadContractsContractConfig<
+      infer TAbi,
+      infer TFunctionName
+    >[]
+  ? ReadContractsContractConfig<TAbi, TFunctionName>[]
+  : ReadContractsContractConfig[]
```

Comment:
Was going to re-use this type logic from `multicall`, but the configuration is subtly different (multicall has `chainId` as a top level option and `readContracts` places `chainId` inside the contract object).